scipy/stats/tests/test_sobol_.py

```diff
@@ -0,0 +1,212 @@
+import scipy.stats.qmc as qmc
+
+
+class TestSobol:
+
+    @pytest.fixture(autouse=True)
+    def setup_sequences(self):
+        # 10 samples for testing. Silent the warning as 10 is not 2**n
+        with pytest.warns(UserWarning):
+            engine_unscrambled_1d = Sobol(1, scramble=False)
+            self.draws_unscrambled_1d = engine_unscrambled_1d.random(10)
+            engine_unscrambled_3d = Sobol(3, scramble=False)
+            self.draws_unscrambled_3d = engine_unscrambled_3d.random(10)
+            engine_scrambled_1d = Sobol(1, scramble=True, seed=12345)
+            self.draws_scrambled_1d = engine_scrambled_1d.random(10)
+            engine_scrambled_3d = Sobol(3, scramble=True, seed=12345)
+            self.draws_scrambled_3d = engine_scrambled_3d.random(10)
+
+    def test_warning(self):
+        with pytest.warns(UserWarning):
+            engine = Sobol(1)
+            engine.random(10)
+
+    def test_random_base2(self):
+        engine = Sobol(1, scramble=False)
+        sample = engine.random_base2(2)
+        assert_array_equal(self.draws_unscrambled_1d[:4],
+                           sample)
+
+        # resampling still having N=2**n
+        sample = engine.random_base2(2)
+        assert_array_equal(self.draws_unscrambled_1d[4:8],
+                           sample)
+
+        # resampling again but leading to N!=2**n
+        with pytest.raises(ValueError, match=r"The balance properties of "
+                                             r"Sobol' points"):
+            engine.random_base2(2)
+
+    def test_raise(self):
+        with pytest.raises(ValueError, match=r"Maximum supported "
+                                             r"dimensionality"):
+            Sobol(Sobol.MAXDIM + 1)
+
+    def test_Unscrambled1DSobol(self):
+        expected = [
+            0, 0.5, 0.75, 0.25, 0.375, 0.875, 0.625, 0.125, 0.1875, 0.6875
+        ]
+        assert_equal(self.draws_unscrambled_1d.shape[0], 10)
+        assert_equal(self.draws_unscrambled_1d.shape[1], 1)
+        assert_array_equal(self.draws_unscrambled_1d.flatten(),
+                           np.array(expected))
+
+    def test_Unscrambled3DSobol(self):
+        expected_dim3 = [0, 0.5, 0.25, 0.75, 0.625, 0.125, 0.875, 0.375,
+                         0.9375, 0.4375]
+        assert_equal(self.draws_unscrambled_3d.shape[0], 10)
+        assert_equal(self.draws_unscrambled_3d.shape[1], 3)
+        assert_array_equal(self.draws_unscrambled_3d[:, 2],
+                           np.array(expected_dim3))
+        assert_array_equal(
+            self.draws_unscrambled_3d[:, 0],
+            self.draws_unscrambled_1d.flatten()
+        )
+
+    def test_Unscrambled3DAsyncSobol(self):
+        engine_unscrambled_3d = Sobol(3, scramble=False)
+        draws = np.vstack([engine_unscrambled_3d.random() for _ in range(10)])
+        assert_array_equal(self.draws_unscrambled_3d, draws)
+
+    def test_UnscrambledFastForwardAndResetSobol(self):
```

Comment:
At least the first part of this test looks like a good candidate to run for every `QMCEngine` subclass. I haven't read through the part after `reset`, but maybe that is a good candidate too.